File: packages/postpin-node/src/resources/pincodes.ts
import type { HttpClient } from "../http.js";
import type { Pincode, RequestOptions, StateSummary } from "../types.js";

/** Pincode directory lookups. */
export class Pincodes {
  constructor(private readonly http: HttpClient) {}

  /**
   * Full details for a single pincode (office, district, state, flags, and a
   * few nearby pincodes in the same district).
   *
   * @example
   * const p = await postpin.pincodes.get("302001"); // Jaipur
   */
  async get(code: string, options: RequestOptions = {}): Promise<Pincode> {
    const { data } = await this.http.request<Pincode>({
      method: "GET",
      path: `/public/pincodes/${encodeURIComponent(code)}`,
      signal: options.signal,
    });
    return data;
  }

  /** List every state/UT with its serviceable-pincode and metro counts. */
  async states(options: RequestOptions = {}): Promise<StateSummary[]> {
    const { data } = await this.http.request<StateSummary[]>({
      method: "GET",
      path: "/public/pincodes/states",
      signal: options.signal,
    });
    return data;
  }
}

File: packages/postpin-python/src/postpin/resources/pincodes.py
from __future__ import annotations

from typing import List, Optional

from .._http import HTTPClient
from ..types import Pincode, StateSummary


class Pincodes:
    def __init__(self, http: HTTPClient) -> None:
        self._http = http

    def get(self, code: str, *, timeout: Optional[float] = None) -> Pincode:
        """Look up a single pincode, including nearby serviceable pincodes."""
        data, _ = self._http.request("GET", f"/public/pincodes/{code}", timeout=timeout)
        return Pincode._from(data)

    def states(self, *, timeout: Optional[float] = None) -> List[StateSummary]:
        """List every state with its serviceable-pincode counts."""
        data, _ = self._http.request("GET", "/public/pincodes/states", timeout=timeout)
        return [StateSummary._from(item) for item in (data or [])]
